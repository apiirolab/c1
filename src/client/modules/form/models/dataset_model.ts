export interface DataSet {
  getValue(name: string): any;
  parseValue(name: string, value: any): void;
  getError(name: string): any;
  getStringValue(name: string): any;
  setStringValue(name: string, value: any): void;
  isExpression(key: string): boolean;
}
