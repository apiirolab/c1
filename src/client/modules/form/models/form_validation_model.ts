export interface FormValidationDao {}

export interface FormValidation {}
