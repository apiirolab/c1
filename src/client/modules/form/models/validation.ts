import i18n from 'es2015-i18n-tag';

export function requiredValidator(value: string | number) {
  return value === '' || value == null ? i18n`This field is required!` : '';
}

export function regExValidator(reg: RegExp, format?: string) {
  return (value: string) => {
    if (reg.exec(value)) {
      return '';
    }
    if (format) {
      return i18n`Expecting format: ${format}`;
    }
    return i18n`Unexpected format`;
  };
}

const intReg = /^(\+|\-)?\d+$/;
function isInt(n: string) {
  return intReg.test(n);
}

const positiveIntReg = /^(\+)?\d+$/;
function isPositiveInt(n: string) {
  return positiveIntReg.test(n);
}

const nonZeroIntReg = /^(\+)?[1-9]\d*$/;
function isNonZeroInt(n: string) {
  return nonZeroIntReg.test(n);
}

const floatReg = /^(\+|\-)?\d*\.?\d+$/;
function isFloat(n: string) {
  return floatReg.test(n);
}

const positiveFloatReg = /^(\+)?\d*\.?\d+$/;
function isPositiveFloat(n: string) {
  return positiveFloatReg.test(n);
}

const nonZeroFloatReg = /^(\+)?[1-9]\d*\.?\d+$/;
function isNonZeroFloat(n: string) {
  return nonZeroFloatReg.test(n);
}

const emailReg = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
function isEmail(n: string) {
  return emailReg.test(n);
}

export function emailValidator(value: string) {
  return value == null || value === '' || isEmail(value) ? '' : i18n`Email has incorrect format!`;
}

export function intPositiveValidator(value: number | string) {
  return value == null || value === '' || isPositiveInt(value.toString())
    ? ''
    : i18n`Expected positive integer value`;
}

export function intValidator(value: number | string) {
  return value == null || value === '' || isInt(value.toString())
    ? ''
    : i18n`Expected integer value`;
}

export function intNonZeroValidator(value: number | string) {
  return value == null || value === '' || isNonZeroInt(value.toString())
    ? ''
    : i18n`Expected non-zero integer value`;
}

export function floatValidator(value: number | string) {
  return value == null || value === '' || isFloat(value.toString())
    ? ''
    : i18n`Expected float value`;
}

export function floatPositiveValidator(value: number | string) {
  return value == null || value === '' || isPositiveFloat(value.toString())
    ? ''
    : i18n`Expected positive float value`;
}

export function floatNonZeroValidator(value: number | string) {
  return value == null || value === '' || isNonZeroFloat(value.toString())
    ? ''
    : i18n`Expected non-zero float value`;
}

export function lengthValidator(length: number, message?: string) {
  return (value: number | string) => {
    return value == null || value === '' || value.toString().length >= length
      ? ''
      : message || i18n`Too short! Minimum ${length} characters`;
  };
}

export function equalityValidator(comparer: () => string | string, message?: string) {
  return (value: string) => {
    let val1 = value ? value.toString() : value.toString();
    let val2 = typeof comparer === 'function' ? comparer() : comparer;

    return val1 === val2 ? '' : message || i18n`Value ${val1} does not match ${val2}`;
  };
}
