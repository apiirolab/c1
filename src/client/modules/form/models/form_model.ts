// import * as shortid from 'shortid';
import * as validations from './validation';

import { IObservableArray } from 'mobx';
import { ISimpleType, types } from 'mobx-state-tree';

import { Yoga } from 'data/yoga';
import { ListValue } from './form_model';
import { FormValidation } from './form_validation_model';

export interface DataSet {
  id?: string;
  addRow(key: string): void;
  removeRow(key: string, index: number): void;
  removeRowData<T>(key: string, data: T): void;
  getList(name: string): ListValue[];
  getValue(name: string): any;
  parseValue(name: string, value: any): void;
  getError(name: string): any;
  getStringValue(name: string): any;
  setStringValue(name: string, value: any): void;
  isExpression(key: string): boolean;
}

export type FormControlProps = {
  formControl: Yoga.FormElement;
  owner: DataSet;
};

function formItemSort(a: Yoga.FormElement, b: Yoga.FormElement) {
  return a.row < b.row
    ? -1
    : a.row > b.row
      ? 1
      : a.column < b.column
        ? -1
        : a.column > b.column
          ? 1
          : 0;
}

function mstTypeFactory(
  desc: Yoga.DataDescriptor,
  lists: Array<{ name: string; items: any[] }>
): ISimpleType<any> {
  switch (desc.type) {
    case 'String':
      return types.string;
    case 'Int':
      return types.number;
    case 'Float':
      return types.number;
    case 'Boolean':
      return types.boolean;
    case 'Object':
      return FormModel.buildMst(desc.descriptors, lists);
    // return For
    case undefined:
      return types.string;
  }
  throw new Error('MST Type not supported: ' + desc.type);
}

export interface ListValue {
  [index: string]: string;
  text: string;
  value: string;
}

let time = Date.now();
let i = 0;

export class FormModel {
  static parseDefault(descriptor: Yoga.DataDescriptor) {
    switch (descriptor.type) {
      case 'Int':
        return parseInt(descriptor.defaultValue || '0', 10) as any;
      case 'Float':
        return parseFloat(descriptor.defaultValue || '0') as any;
      case 'Boolean':
        return (descriptor.defaultValue === 'true' || descriptor.defaultValue === 'True') as any;
      case 'Date':
        return new Date(descriptor.defaultValue);
    }
    return descriptor.defaultValue;
  }

  static buildMst(
    descriptors: Yoga.DataDescriptor[],
    lists?: Array<{ name: string; items: any[] }>
  ) {
    const descriptorMap: {
      [index: string]: Yoga.DataDescriptor;
    } = {};
    for (let d of descriptors) {
      descriptorMap[d.name] = d;
    }

    // prepare model and views
    const mstDefinition: { [index: string]: any } = {};
    const viewDefinition = (self: any) => {
      const view = {
        getDescriptor(key: string) {
          if (!descriptorMap[key]) {
            throw new Error('Descriptor does not exist: ' + key);
          }
          return descriptorMap[key];
        },
        isExpression(key: string) {
          return !!(key && this.getDescriptor(key).expression);
        },
        getValue(key: string) {
          if (self[key] == null) {
            return this.getDescriptor(key).defaultValue;
          }
          return self[key];
        },
        getList(key: string): ListValue[] {
          return self[key];
        },
        getStringValue(key: string) {
          // expressions are evaluated on the fly
          if (this.getDescriptor(key).expression) {
            // @ts-ignore
            return self[key];
          }
          return self[key + '_str'] || this.getDescriptor(key).defaultValue || '';
        },
        getError(key: string) {
          return self[key + '_error'];
        }
      };

      // add expressions views
      // expression are dynamic formulas evaluated from script

      for (let desc of descriptors) {
        // expressions do not need state tree entry they are evaluated automatically
        if (desc.expression) {
          (view as any).__defineGetter__(desc.name, function() {
            // tslint:disable-next-line:no-eval
            return eval(desc.expression);
          });
        }
      }

      if (lists) {
        for (let list of lists) {
          (view as any).__defineGetter__(list.name, function() {
            return list.items;
          });
        }
      }

      return view;
    };

    // add mst nodes for non expression fields
    // expressions do not need custom nodes and are handled by views
    for (let desc of descriptors) {
      // expressions do not need state tree entry they are evaluated automatically
      if (desc.isArray) {
        mstDefinition[desc.name] = types.array(mstTypeFactory(desc, lists));
      } else if (desc.type === 'Id') {
        mstDefinition[desc.name] = types.optional(types.identifier, () => (time + i++).toString()); // shortid.generate());
      } else if (!desc.expression) {
        mstDefinition[desc.name] = desc.defaultValue
          ? FormModel.parseDefault(desc)
          : types.maybe(mstTypeFactory(desc, lists));
        mstDefinition[desc.name + '_str'] = types.maybe(types.string);
        mstDefinition[desc.name + '_error'] = types.maybe(types.string);
      }
    }

    // add generic method

    // build tree
    const mst = types
      .model('Store', mstDefinition)
      .views(viewDefinition)
      .actions(self => ({
        addRow(key: string) {
          (self[key] as any[]).push({});
        },
        removeRow(key: string, index: number) {
          (self[key] as any[]).splice(index);
        },
        removeRowData<T>(key: string, data: T) {
          (self[key] as IObservableArray<T>).remove(data);
        },
        parseValue(key: string, value: any) {
          const descriptor = descriptorMap[key];
          switch (descriptor.type) {
            case 'Int':
              self[key] = parseInt(value || 0, 10) as any;
              break;
            case 'Float':
              self[key] = parseFloat(value || 0) as any;
              break;
            case 'Boolean':
              self[key] = (value === true || value === 'true' || value === 'True') as any;
              break;
            default:
              self[key] = value;
          }
        },
        setError(key: string, error: string) {
          (self as any)[key + '_error'] = error;
        }
      }))
      .actions(self => ({
        setStringValue(key: string, value: any) {
          self[key + '_str'] = value;

          const descriptor = descriptorMap[key];
          let error = '';
          switch (descriptor.type) {
            case 'Int':
              error = validations.intValidator(value);
              break;
            case 'Float':
              error = validations.floatValidator(value);
              break;
          }

          if (!error) {
            self.parseValue(key, value);
          }
          self.setError(key, error);
        }
      }));

    return mst;
  }

  static initStrings(data: any, dataArray: Array<{ name: string; value: any }>) {
    for (let item of dataArray) {
      data[item.name] = item.value;
      if (Array.isArray(item.value)) {
        data[item.name] = [];
        for (let m of item.value) {
          const n = {};
          data[item.name].push(n);
          FormModel.initStrings(n, m);
        }
      } else {
        data[item.name + '_str'] = item.value ? item.value.toString() : '';
      }
    }

    return data;
  }

  static buildMstModel(
    descriptors: Yoga.DataDescriptor[],
    dataArray: Array<{ name: string; value: any }>,
    lists?: Array<{ name: string; items: any[] }>
  ): DataSet {
    // data initialisation
    const data: any = {};
    FormModel.initStrings(data, dataArray);

    const mst = FormModel.buildMst(descriptors, lists);
    return mst.create(data);
  }

  id: string;
  name: string;
  description: string;
  elements: Yoga.FormElement[];
  validations: FormValidation[];

  constructor(form: Yoga.Form) {
    this.id = form.id;
    this.name = form.name;
    this.description = form.description;
    this.validations = form.validations;

    // add and sort elements
    this.elements = form.elements;
    this.elements.sort(formItemSort);
  }
}
