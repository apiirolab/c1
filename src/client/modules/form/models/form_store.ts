import { IObservableArray, toJS } from 'mobx';
import { types } from 'mobx-state-tree';
import { UndoManager } from 'mst-middlewares';

import { QueryTypes } from 'data/client';
import { DataDescriptor, DataSet } from './form_model';

export type IValidator = (input: string) => string;

export type IFormStore = typeof FormStore.Type;

export type ValidationResult = {
  valid: number;
  invalid: number;
  total: number;
};

function strip(obj: any) {
  if (typeof obj === 'object') {
    if (obj.errors) {
      delete obj.errors;
      delete obj.validators;
      delete obj.strings;
      delete obj.descriptors;
      delete obj.arrays;
      delete obj.objects;
    }

    for (let key of Object.getOwnPropertyNames(obj)) {
      let item = obj[key];
      strip(item);
    }
  }

  if (Array.isArray(obj)) {
    for (let item of obj) {
      strip(item);
    }
  }
  return obj;
}

export function isRequired(descriptor: QueryTypes.DataDescriptor) {
  return (
    descriptor.validators &&
    descriptor.validators.length &&
    descriptor.validators.some(d => d.name === QueryTypes.ValidatorFunction.RequiredValidator)
  );
}

export interface ListValue {
  [index: string]: string;
  text: string;
  value: string;
}

export const FormStore = types
  .model({
    dirty: false,
    errors: types.map(types.string),
    strings: types.map(types.string)
  })
  .volatile(self => ({
    arrays: [],
    objects: [],
    // tslint:disable-next-line:no-object-literal-type-assertion
    validators: {} as { [name: string]: IValidator[] },
    // tslint:disable-next-line:no-object-literal-type-assertion
    descriptors: {} as { [index: string]: DataDescriptor },
    helpers: {
      getDescriptor(key: string) {
        if (!self.descriptors[key]) {
          throw new Error('Descriptor does not exist: ' + key);
        }
        return self.descriptors[key];
      },
      isExpression(key: string) {
        return !!(key && this.getDescriptor(key).expression);
      }
    }
  }))
  .actions(function(self) {
    // let validators: values: { [name: string]: IValidator[] } = new Map<string, IValidator[]>();

    function setValue(key: string, value: any) {
      if ((self as any)[key] !== value) {
        (self as any)[key] = value;
        self.dirty = true;
      }
    }

    return {
      saved() {
        self.dirty = false;
      },
      toJS() {
        return strip(toJS(self));
      },

      getList(key: string): ListValue[] {
        return (self as any)[key];
      },
      validateField(name: string, value: string) {
        if (!self.validators[name]) {
          return true;
        }

        self.errors.set(name, '');

        for (let v of self.validators[name]) {
          if (v) {
            let result = v(value);
            if (result) {
              self.errors.set(name, result);
              return false;
            }
          }
        }

        return true;
      },
      // setValidators(values: { [name: string]: IValidator[] }) {
      //   for (let key of Object.getOwnPropertyNames(values)) {
      //     self.validators[key] = values[key];
      //   }
      // },
      setError(item: string, value: string) {
        self.errors.set(item, value);
        if (value) {
          self.errors.set(item, value);
        } else {
          self.errors.delete(item);
        }
      },
      setItem(item: string, value: string): void {
        (self as any)[item] = value;
        (self as IFormStore).validateField(item, value);
      },
      addRow(key: string) {
        ((self as any)[key] as any[]).push({});
      },
      removeRow(key: string, index: number) {
        ((self as any)[key] as any[]).splice(index);
      },
      removeRowData<T>(key: string, data: T) {
        ((self as any)[key] as IObservableArray<T>).remove(data);
      },
      parseValue(key: string, value: any) {
        const descriptor = self.descriptors[key];
        switch (descriptor.type) {
          case 'Int':
            setValue(key, parseInt(value || 0, 10));
            break;
          case 'Float':
            setValue(key, parseFloat(value || 0));
            break;
          case 'Boolean':
            setValue(key, value === true || value === 'true' || value === 'True');
            break;
          default:
            setValue(key, value);
        }
      }
    };
  })
  .actions(self => ({
    validateWithReport(): ValidationResult {
      let total = 0;
      let valid = 0;

      // validate self

      for (let key of Object.getOwnPropertyNames(self.validators)) {
        total += 1;
        if (self.validateField(key, (self as any)[key])) {
          valid += 1;
        }
      }

      // validate objects

      for (let key of self.objects) {
        let obj = (self as IFormStore).getValue(key) as DataSet;
        let result = obj.validateWithReport();
        total += result.total;
        valid = result.valid;
      }

      // validate arrays

      for (let key of self.arrays) {
        let obj = (self as IFormStore).getValue(key) as DataSet[];

        for (let row of obj) {
          let result = row.validateWithReport();
          total += result.total;
          valid = result.valid;
        }
      }

      // create report

      return {
        total,
        valid,
        invalid: total - valid
      };
    }
  }))
  .actions(self => ({
    validate() {
      return self.validateWithReport().invalid === 0;
    },
    setStringValue(key: string, value: any, validate = true) {
      self.strings.set(key, value);

      if (validate) {
        let isOK = self.validateField(key, value);

        if (isOK) {
          self.parseValue(key, value);
        }
      }
      // self.setError(key, error);
    }
  }))
  .views(self => ({
    getValue(item: string): any {
      return (self as any)[item];
    },
    getStringValue(key: string): string {
      const descriptor = self.helpers.getDescriptor(key);
      // expressions are evaluated on the fly
      if (descriptor.expression) {
        // @ts-ignore
        return self[key];
      }

      if (descriptor.type === QueryTypes.DataType.Object) {
        return '<ERROR! Rendering Object as TEXT>';
      }

      // init
      if (!self.strings.has(key)) {
        self.setStringValue(
          key,
          (self as any)[key] == null ? '' : (self as any)[key].toString(),
          false
        );
      }
      return self.strings.get(key); // || this.getDescriptor(key).defaultValue || '';
    },
    getError(item: string): string {
      return self.errors.get(item);
    }
  }));

// export const FormStore = types
//   .model({
//     errors: types.map(types.string)
//   })
//   .actions(self => ({
//     setItem(item: string, value: string): void {
//       (self as any)[item] = value;
//     },
//     getItem(item: string): string {
//       return (self as any)[item];
//     },
//     getError(item: string): string {
//       return self.errors.get(item);
//     }
//   }));
