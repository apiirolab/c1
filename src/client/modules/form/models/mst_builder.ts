import { Instance, types } from 'mobx-state-tree';
import { UndoManager } from 'mst-middlewares';

import { DataSetNode } from './dataset_model';
import { FormStore } from './form_store';
import { safeEval } from './form_utils';

let time = Date.now();
let i = 0;
function shortId() {
  return (time + i++).toString();
}

function mstTypeFactory(desc: DataSetNode): any {
  switch (desc.type) {
    case 'id':
      return types.optional(types.identifier, shortId);
    case 'expression':
      return undefined;
    case 'array':
      return types.optional(
        types.array(types.optional(mstTypeFactory(desc.items), desc.items.defaultValue)),
        desc.default || []
      );
    case 'string':
      return types.optional(types.union(types.string, types.undefined), desc.default);
    case 'integer':
      return types.optional(types.union(types.number, types.string, types.undefined), desc.default);
    case 'number':
      return types.optional(types.union(types.number, types.string, types.undefined), desc.default);
    case 'date':
      return types.optional(types.union(types.Date, types.string, types.undefined), desc.default);
    case 'boolean':
      return types.optional(
        types.union(types.boolean, types.string, types.undefined),
        desc.default
      );
    case 'object':
      return types.optional(buildStore(desc), {});
    case undefined:
      return types.string;
  }
  throw new Error('MST Type not supported: ' + desc.type);
}

export function buildStore(store: DataSetNode) {
  // prepare model and views

  const mstDefinition: { [index: string]: any } = {};

  /* =========================================================
    EXPRESSIONS
    ======================================================== */
  const properties = Object.getOwnPropertyNames(store.properties);

  const viewDefinition = () => {
    const view = {};

    for (let key of properties) {
      let node = store.properties[key];
      // expressions do not need state tree entry they are evaluated automatically
      if (node.type === 'expression') {
        (view as any).__defineGetter__(key, function() {
          // @ts-ignore
          return safeEval(this, node.default);
        });
      }
    }

    return view;
  };

  /* =========================================================
      MST Nodes
     ======================================================== */

  for (let key of properties) {
    let node = store.properties[key];
    let definition = mstTypeFactory(node);
    if (definition) {
      mstDefinition[key] = types.maybe(definition);
    }
  }

  if (store.parent == null) {
    mstDefinition.history = types.optional(UndoManager, {});
  }

  // build tree
  const mst = FormStore.named('FormStore')
    .props(mstDefinition)
    .views(viewDefinition)
    .actions(() => ({
      getSchema(key: string) {
        return key ? store.properties[key] : store;
      }
    }));

  return mst;
}
