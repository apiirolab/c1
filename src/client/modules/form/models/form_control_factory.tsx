import * as React from 'react';

import { Button } from 'semantic-ui-react';
import { DataSet } from './form_model';
import { CheckboxView } from '../views/checkbox_view';
import { FormulaView } from '../views/formula_view';
import { InputView } from '../views/input_view';
import { RadioView } from '../views/radio_view';
import { RepeaterView } from '../views/repeater_view';
import { SelectView } from '../views/select_view';

export function renderControl(
  control: Corpix.Entities.FormElement,
  dataSet: DataSet,
  handlers: { [index: string]: () => void }
) {
  const formElement = control as Corpix.Entities.FormControl;

  if (dataSet.isExpression(formElement.source)) {
    return <FormulaView owner={dataSet} formControl={formElement} />;
  }

  switch (formElement.control) {
    case 'input':
      return <InputView owner={dataSet} formControl={formElement} />;
    case 'select':
      return <SelectView owner={dataSet} formControl={formElement} />;
    case 'checkbox':
      return <CheckboxView owner={dataSet} formControl={formElement} />;
    case 'radio':
      return <RadioView owner={dataSet} formControl={formElement} />;
    case 'repeater':
      return <RepeaterView owner={dataSet} formControl={formElement} />;
    case 'deleteButton':
      return <Button icon="trash" color="red" onClick={handlers.delete} />;
  }

  throw new Error('Not implemented: ' + formElement.control);
}
