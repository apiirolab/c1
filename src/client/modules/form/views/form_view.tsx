import * as React from 'react';

import { Form, SemanticWIDTHSNUMBER } from 'semantic-ui-react';

import { groupByArray } from 'data/helpers';

import { QueryTypes } from 'data/client';
import { renderControl } from '../models/form_control_factory';
import { DataSet, FormElement } from '../models/form_model';

interface IFieldOwner {
  elements?: QueryTypes.Form_Elements[];
}

interface Props {
  data: DataSet;
  form: IFieldOwner;
  report?: boolean;
  handlers?: { [index: string]: () => void };
}

export class FormView extends React.Component<Props> {
  lastRow = -1;
  lastColumn = -1;

  renderColumn(control: FormElement) {
    if (control.row !== this.lastRow) {
      this.lastRow = control.row;
      this.lastColumn = 0;
    }
    // we initialise all columns and add missing ones in between
    let columns = [];
    const formControl = control;

    if (formControl.control === 'DeleteButton') {
      formControl.label = '\xA0';
    }

    // insert missing start column
    if (control.column > this.lastColumn) {
      columns.push(
        <Form.Field
          key={this.lastColumn}
          width={(control.column - this.lastColumn) as SemanticWIDTHSNUMBER}
        >
          &nbsp;
        </Form.Field>
      );
    }

    columns.push(
      <Form.Field
        key={control.column}
        width={formControl.inline ? undefined : (control.width as SemanticWIDTHSNUMBER)}
        inline={formControl.inline}
      >
        {formControl.label &&
          formControl.control !== 'Checkbox' &&
          formControl.label !== 'Radio' && (
            <label htmlFor={(formControl.source && formControl.source.name) || undefined}>
              {formControl.label}
            </label>
          )}
        {renderControl(control, this.props.data, this.props.handlers)}
      </Form.Field>
    );

    this.lastColumn = control.column + control.width;
    return columns;
  }

  render() {
    this.lastColumn = 0;
    this.lastRow = 0;

    let row: { key: number; values: FormElement[] };
    const rows = groupByArray(this.props.form.elements, 'row');

    return (
      <div className={'ui form' + (this.props.report ? ' report' : '')}>
        <For each="row" of={rows}>
          <Form.Group key={row.key}>
            {row.values.map(element => this.renderColumn(element))}
          </Form.Group>
        </For>
      </div>
    );
  }
}
