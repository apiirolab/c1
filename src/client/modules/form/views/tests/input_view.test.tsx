import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Segment } from 'semantic-ui-react';

import { create } from '../../../../../shared/test_data';
import { FormModel } from '../../models/form_model';
import { FormView } from '../form_view';

describe('Form', () => {
  const descriptors = [
    create.descriptor({ name: 'owner.personal.name' }),
    create.descriptor({ name: 'owner.personal.age', type: 'float' }),
    create.descriptor({
      name: 'younger',
      type: 'int',
      expression: `this['owner.personal.age'] - 10`
    }),
    create.descriptor({
      name: 'older',
      type: 'int',
      expression: `this['owner.personal.age'] + 10`
    })
  ];

  const controlData = [
    { name: 'owner.personal.name', value: 'Tomas' },
    {
      name: 'owner.personal.age',
      value: 33
    }
  ];

  const dataSet = FormModel.buildMstModel(descriptors, controlData);

  storyOf(
    'Viewer',
    {
      get component() {
        const form = new FormModel(
          create.formDao({
            elements: [
              {
                id: '1',
                row: 0,
                column: 0,
                width: 16,
                control: 'input',
                controlProps: {
                  label: 'Name'
                },
                label: 'Mimo',
                source: 'owner.personal.name'
              },
              {
                id: '2',
                row: 1,
                column: 1,
                width: 7,
                control: 'input',
                source: 'owner.personal.age',
                label: 'Age: ',
                inline: true
              },
              {
                id: '3',
                row: 1,
                column: 10,
                width: 2,
                control: 'input',
                source: 'younger',
                label: 'Younger'
              },
              {
                id: '4',
                row: 1,
                column: 13,
                width: 2,
                control: 'input',
                source: 'older',
                label: 'Older'
              }
            ]
          })
        );

        // just another notation
        return (
          <Segment className="ui form">
            <FormView form={form} data={dataSet} />
          </Segment>
        );
      }
    },
    data => {
      it('renders correctly', () => {
        const component = renderer.create(data.component);
        expect(component).toMatchSnapshot();
      });


      it('changes value and all related formulas', () => {
        const component = renderer.create(data.component);
        const root = component.root;
        const age = root.findByProps({ name: 'owner.personal.age' });
        age.props.onChange({ target: { value: '40' } });
        expect(component).toMatchSnapshot();
      });
    }
  );
});
