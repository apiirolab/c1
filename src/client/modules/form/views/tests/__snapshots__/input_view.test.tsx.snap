// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Input changes value and all related formulas 1`] = `
<div
  className="ui segment ui form"
>
  <div
    className="ui form"
  >
    <div
      className="fields"
    >
      <div
        className="fourteen wide field"
      >
        <label
          htmlFor="owner.personal.name"
        >
          Mimo
        </label>
        <div
          className="ui labeled input"
        >
          <div
            className="ui label label"
            onClick={[Function]}
          >
            Name
          </div>
          <input
            name="owner.personal.name"
            onChange={[Function]}
            type="text"
            value="Tomas"
          />
        </div>
      </div>
      <div
        className="two wide field"
      >
        <label
          htmlFor="older"
        >
          Older
        </label>
        <div
          className="ui input"
        >
          <div
            className="sc-bdVaJa iDwIel"
          >
            50
          </div>
        </div>
      </div>
    </div>
    <div
      className="fields"
    >
      <div
        className="one wide field"
      >
         
      </div>
      <div
        className="inline field"
      >
        <label
          htmlFor="owner.personal.age"
        >
          Age: 
        </label>
        <div
          className="ui input"
        >
          <input
            name="owner.personal.age"
            onChange={[Function]}
            type="text"
            value="40"
          />
        </div>
      </div>
      <div
        className="two wide field"
      >
         
      </div>
      <div
        className="inline field"
      >
        <label
          htmlFor="younger"
        >
          Younger:
        </label>
        <div
          className="ui input"
        >
          <div
            className="sc-bdVaJa iDwIel"
          >
            20
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Form Input renders correctly 1`] = `
<div
  className="ui segment ui form"
>
  <div
    className="ui form"
  >
    <div
      className="fields"
    >
      <div
        className="fourteen wide field"
      >
        <label
          htmlFor="owner.personal.name"
        >
          Mimo
        </label>
        <div
          className="ui labeled input"
        >
          <div
            className="ui label label"
            onClick={[Function]}
          >
            Name
          </div>
          <input
            name="owner.personal.name"
            onChange={[Function]}
            type="text"
            value="Tomas"
          />
        </div>
      </div>
      <div
        className="two wide field"
      >
        <label
          htmlFor="older"
        >
          Older
        </label>
        <div
          className="ui input"
        >
          <div
            className="sc-bdVaJa iDwIel"
          >
            43
          </div>
        </div>
      </div>
    </div>
    <div
      className="fields"
    >
      <div
        className="one wide field"
      >
         
      </div>
      <div
        className="inline field"
      >
        <label
          htmlFor="owner.personal.age"
        >
          Age: 
        </label>
        <div
          className="ui input"
        >
          <input
            name="owner.personal.age"
            onChange={[Function]}
            type="text"
            value="33"
          />
        </div>
      </div>
      <div
        className="two wide field"
      >
         
      </div>
      <div
        className="inline field"
      >
        <label
          htmlFor="younger"
        >
          Younger:
        </label>
        <div
          className="ui input"
        >
          <div
            className="sc-bdVaJa iDwIel"
          >
            13
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
